JoomGalleryfriends/JG4-dev
# Tus upload service

## What is this service for
The TUS server service will be used to upload multiple files in a Drag'N'Drop form input field.
The core of this service is an upload server implementing the [TUS resumable upload protocol](https://tus.io/protocols/resumable-upload.html). The TUS server provides a backend to receive files from any upload client implementing the TUS protocol.

### Backend implementataion
The TUS server in the backend was created using https://github.com/Orajo/zf2-tus-server. The code was modified to use [Joomla\CMS\Application\CMSApplication](https://api.joomla.org/cms-3/classes/Joomla.CMS.Application.CMSApplication.html) instead of the laminas framework to create the http responses.

### Frontend implementation
The uploading form in the fronted contains the upload form field [Uppy](https://uppy.io/examples/dashboard/). This is the official frontend upload client from the developer of the TUS protocol.

## How does it work
Upload requests are sent from the frontend client via XHR request to the URL `/administrator/index.php?option=com_joomgallery&task=images.tusupload`. This will start the TUS server handling the incoming requests and saves the approaching data to the Joomla temp folder. The uploaded files are saved with the Upload-ID (UUID) as filename.

## How to use this service
The service gets loaded automatically when you send a request with the task `task=images.tusupload` to the JoomGallery component in the Joomla backend.
But you can load the service anywhere in Joomla by doing

```
$component = Factory::getApplication()->bootComponent('com_joomgallery');  // boot JoomGallery component
$this->component->createTusServer();  // instantiate service
$server = $this->component->getTusServer();  // load the service into your scope
$server->process(true);  // run tus server
```

When instantiating the service you could add a parameter to the createTusServer() method specifying the folder where uploaded files should be stored. Default upload folder is the Joomla temporary (temp) forlder.

## How to test the service
Make sure "Debug System" is "Yes" in Joomla Global configuration.

Navigate to Components-JoomGallery-Images in the Joomla backend.
![grafik](https://user-images.githubusercontent.com/39154009/222983471-107ecd34-9062-45b1-ad9c-7acdef240fb0.png)
The "New"-Button now has a ned sub-button called "Multiple New". This button brings you to the upload from with an uppy form field. Start uploading files using this uppy form field.
After upload a button with debug information is added to the thumb of theuploaded file:
![grafik](https://user-images.githubusercontent.com/39154009/222984171-8e6875d8-654b-434d-93e2-a3917ba36e6d.png)
Read out the Upload-ID (UUID) in the Debug information and check the following:

- A file is created on the server in the upload folder (Joomla temp folder) with the UUID as filename.
- A second file is created called <UUID>.info containing different informations of the uploaded file.
- Rename the created file (with UUID as filename) and add the file extension to the filename.
- Open the renamed file and check its content.

### Normal file upload
You can upload any files up to a size of 250MB. Try uploading different file types. The files gets uploaded into the temp folder of Joomla. They are stored as files with filename <UUID>. You can find the Upload-ID (UUID) in the Debug information of the file after successful upload.

### Large files
Test, if you are able to upload large files. Test if you can pause and resume the upload.

### Add file information
By using the edit button, you can add some file specific information to the upload. Add some information and check if the added info is stored successfully in the <UUID>.info file in the upload folder.
![grafik](https://user-images.githubusercontent.com/39154009/222983754-85375cd7-0da6-4b64-ac6a-2ff153838d8a.png)

### Interrupted internet connection
If your connection to the server is interrupted during upload, the upload stopps and can be resumed as soon as you have internt connection again.

### Different servers
If possible try to use different servers (local, hoster, Windows, Linux, ...).

### Different browsers
Try to test the upload with different browsers.
